### PR TITLE
fix(service): change default service name

### LINKED ISSUE #273: [BUG] default service name must be docker not dockerd
### Formula commit hash / release tag

https://github.com/saltstack-formulas/docker-formula/releases/tag/v1.1.1

### Versions reports (master & minion)

<details><summary>master</summary>

```bash
# salt --versions-report
Salt Version:
          Salt: 3002.2

Dependency Versions:
          cffi: Not Installed
      cherrypy: Not Installed
      dateutil: 2.5.3
     docker-py: Not Installed
         gitdb: 2.0.0
     gitpython: 2.1.1
        Jinja2: 2.9.4
       libgit2: Not Installed
      M2Crypto: Not Installed
          Mako: Not Installed
       msgpack: 0.6.2
  msgpack-pure: Not Installed
  mysql-python: Not Installed
     pycparser: Not Installed
      pycrypto: Not Installed
  pycryptodome: 3.6.1
        pygit2: Not Installed
        Python: 3.5.3 (default, Jul  9 2020, 13:00:10)
  python-gnupg: Not Installed
        PyYAML: 3.12
         PyZMQ: 17.1.2
         smmap: 2.0.1
       timelib: Not Installed
       Tornado: 4.5.3
           ZMQ: 4.2.1

System Versions:
          dist: debian 9 stretch
        locale: UTF-8
       machine: x86_64
       release: 4.9.0-13-amd64
        system: Linux
       version: Debian GNU/Linux 9 stretch
```

</details>

<details><summary>minion</summary>

```bash
# salt-minion --version-report
Salt Version:
          Salt: 3002.5

Dependency Versions:
          cffi: Not Installed
      cherrypy: Not Installed
      dateutil: 2.6.1
     docker-py: 4.4.4
         gitdb: Not Installed
     gitpython: Not Installed
        Jinja2: 2.10.1
       libgit2: Not Installed
      M2Crypto: 0.35.2
          Mako: Not Installed
       msgpack: 0.6.2
  msgpack-pure: Not Installed
  mysql-python: Not Installed
     pycparser: Not Installed
      pycrypto: Not Installed
  pycryptodome: Not Installed
        pygit2: Not Installed
        Python: 3.6.8 (default, Aug 24 2020, 17:57:11)
  python-gnupg: Not Installed
        PyYAML: 3.12
         PyZMQ: 19.0.0
         smmap: Not Installed
       timelib: Not Installed
       Tornado: 4.5.3
           ZMQ: 4.3.4

System Versions:
          dist: centos 8 n/a
        locale: UTF-8
       machine: x86_64
       release: 4.18.0-240.1.1.el8_3.x86_64
        system: Linux
       version: CentOS Linux 8 n/a
```
</details>

### Pillar / config used

```yaml
docker:
  pkg:
    docker:
      use_upstream: repo
```

---

### Describe the bug
Default service name in current official docker packages for CentOS / Debian is `docker` not `dockerd` but formula default is `dockerd`.

https://github.com/saltstack-formulas/docker-formula/blob/8835a9f1977e3a569bca705a6d1063e1c3f115b9/docker/defaults.yaml#L40-L43

On CentOS default service status after package installed is `Loaded: loaded (/usr/lib/systemd/system/docker.service; disabled; vendor preset: disabled)`
As a result service is not started, not enabled after formula applied.

On Debian 10 default  service status after package installed is `Loaded: loaded (/lib/systemd/system/docker.service; enabled; vendor preset: enabled)` 
Therefore this issue can be missed while using formula with Debian.

Also although service `dockerd` is clearly missing formula fails silently due unnecessary `onlyif` condition
https://github.com/saltstack-formulas/docker-formula/blob/8835a9f1977e3a569bca705a6d1063e1c3f115b9/docker/software/service/running.sls#L47

It is not clear why this condition is added in general. It's only masking issues like this.

### Steps to reproduce the bug
On CentOS 8

```bash
# install docker with formula using upstream repository
#  docker:pkg:docker:use_upstream: repo
$ salt 'minion-centos8' state.sls docker
# check service
$ salt 'minion-centos8' service.missing docker
minion-centos8:
    False
$ salt 'minion-centos8' service.missing dockerd
minion-centos8:
    True
```

### Expected behaviour
After formula is applied to the minion, `docker` service is running and operational.

### Attempts to fix the bug
Issue can be fixed locally by adding correct service name to pillars

```yaml
docker:
  pkg:
    docker:
      use_upstream: repo
      service:
        name: docker
```
But defaults must be fixed too.

### Additional context
I think this change can be considered as BREAKING. After this change duplicate service will be created on systems where `archive` method is used. Therefor I'm creating issue to discuss.

###

### PR DESCRIPTION
**Pull request by hatifnatt**
### PR progress checklist (to be filled in by reviewers)

- [ ] Changes to documentation are appropriate (or tick if not required)
- [ ] Changes to tests are appropriate (or tick if not required)
- [ ] Reviews completed

---

#### Primary type

- [ ] `[build]`    Changes related to the build system
- [ ] `[chore]`    Changes to the build process or auxiliary tools and libraries such as documentation generation
- [ ] `[ci]`       Changes to the continuous integration configuration
- [ ] `[feat]`     A new feature
- [x] `[fix]`      A bug fix
- [ ] `[perf]`     A code change that improves performance
- [ ] `[refactor]` A code change that neither fixes a bug nor adds a feature
- [ ] `[revert]`   A change used to revert a previous commit
- [ ] `[style]`    Changes that do not affect the meaning of the code (white-space, formatting, missing semi-colons, etc.)

#### Secondary type

- [x] `[docs]`     Documentation changes
- [ ] `[test]`     Adding missing or correcting existing tests

### Does this PR introduce a `BREAKING CHANGE`?

Yes.

BREAKING CHANGE: due changes in default service name, on systems where 'archive' installation method is used, duplicate service will be created. This can be avoided by updating pillar with 'docker:pkg:docker:service:name: dockerd'. Due fact that 'archive' method is default this change may affect a large number of users

### Related issues and/or pull requests

Closes #273 

### Describe the changes you're proposing

Docker daemon service name in official Docker packages is 'docker' but formula uses 'dockerd'. Due to these differences docker service won't start after installation if `package` installation method is selected - `docker:pkg:docker:use_upstream: repo` .

Fix documentation link in systemd service

`docker/software/archive/install.sls`
Docker daemon binary in official Docker archive distribution is always 'dockerd' no reason to substitute service name

`defaults.yaml`
'null' from yaml is rendering as 'None' by Jinja therefore, systemd displays a warning in the log, replace with empty string to prevent this issue

### Pillar / config required to test the proposed changes

```yaml
docker:
  pkg:
    docker:
      use_upstream: repo
```

### Debug log showing how the proposed changes work
```
2021-03-10 17:24:59,564 [salt.template    :122 ][DEBUG   ][2654] Rendered data from file: /var/cache/salt/minion/files/base/docker/software/service/running.sls:
# -*- coding: utf-8 -*-
# vim: ft=sls

include:
  - docker.software.package.install
  - docker.software.config.file

docker-software-service-running-unmasked:
  service.unmasked:
    - name: docker
    - onlyif: systemctl list-unit-files | grep docker >/dev/null 2>&1
    - require_in:
      - service: docker-software-service-running-docker

docker-software-service-running-docker:
  service.running:
    - name: docker
    - enable: True

docker-software-service-running-docker-fail-notify:
  test.fail_without_changes:
    - comment: |
        Formula is trying to start 'docker' service
        but failed, is it a correct name for Docker service in your OS?

        In certain circumstances the docker service will not start.
        Your kernel is missing some modules, or not in ideal state.
        See https://github.com/moby/moby/blob/master/contrib/check-config.sh
        * Rebooting your host is recommended!
    - onfail:
      - service: docker-software-service-running-docker
  service.enabled:
    - onlyif: True
    - name: docker
    - onfail:
      - service: docker-software-service-running-docker
```

### Documentation checklist

- [ ] Updated the `README` (e.g. `Available states`).
- [x] Updated `pillar.example`.

### Testing checklist

- [ ] Included in Kitchen (i.e. under `state_top`).
- [ ] Covered by new/existing tests (e.g. InSpec, Serverspec, etc.).
- [ ] Updated the relevant test pillar.

### Additional context

Better merge this after #276

### CODE REVIEW

**Comment by myii** [comment]
@hatifnatt Looks like you're going to need to modify this test as well:

https://github.com/saltstack-formulas/docker-formula/blob/3f8513119a5ce32dcb8413f983e582fea5defa28/test/integration/archive/controls/service.rb#L6

**Comment by hatifnatt** [comment]
@myii good point, I'm still not using Kitchen locally
Does this need to be uncommented too?
https://github.com/saltstack-formulas/docker-formula/blob/3f8513119a5ce32dcb8413f983e582fea5defa28/test/integration/archive/controls/service.rb#L9
Title is stating 'running and enabled' but it seems like `running`  is not  actually tested.
https://github.com/saltstack-formulas/docker-formula/blob/3f8513119a5ce32dcb8413f983e582fea5defa28/test/integration/archive/controls/service.rb#L4

**Comment by myii** [comment]
> Does this need to be uncommented too?
> 
> https://github.com/saltstack-formulas/docker-formula/blob/3f8513119a5ce32dcb8413f983e582fea5defa28/test/integration/archive/controls/service.rb#L9
> 
> Title is stating 'running and enabled' but it seems like `running` is not actually tested.
> 
> https://github.com/saltstack-formulas/docker-formula/blob/3f8513119a5ce32dcb8413f983e582fea5defa28/test/integration/archive/controls/service.rb#L4

@hatifnatt Ideally, yes -- but there's probably a reason why @noelmcloughlin disabled it, so let's leave it for this particular PR.  I suppose some of the instances will fail due to some container-based issues.

In fact, this block has also been disabled and would need to be investigated:

https://github.com/saltstack-formulas/docker-formula/blob/3f8513119a5ce32dcb8413f983e582fea5defa28/test/integration/package/controls/service.rb#L6-L9

**Review by myii — APPROVED** [review]
Thanks, @hatifnatt.  Since this involves a `BREAKING CHANGE`, I'll leave this open for a second opinion, probably from @noelmcloughlin.

**Comment by noelmcloughlin** [comment]
thanks @hatifnatt 
Can we discuss service name at https://github.com/saltstack-formulas/docker-formula/issues/273#issuecomment-796150227

**Comment by noelmcloughlin** [comment]
> Title is stating 'running and enabled' but it seems like `running` is not actually tested.

@myii the `service.running` state did not work in Travis CI/CD so this was fudge.

**Comment by myii** [comment]
> > Title is stating 'running and enabled' but it seems like `running` is not actually tested.
> 
> @myii the **docker** `service.running` state did not work in Travis CI/CD so this was fudge. So yes, that was the reason those tests were disabled.

@noelmcloughlin Thanks for the response.  I'm intrigued... let me try a run here in GitHub Actions to see what happens.  I'll let you know shortly.

**Comment by noelmcloughlin** [comment]
Service name is (probably) `docker` in linux native package format, but is (definitely) `dockerd` with official archive releases. https://github.com/saltstack-formulas/docker-formula/issues/273#issuecomment-796182326

**Comment by noelmcloughlin** [comment]
> I'm intrigued... let me try a run here in GitHub Actions

In Travis CI/CD we used docker driver.  In some linux distributions `docker` service has issues with `iptables/kernel` and reboot is necessary.  This is all I recollect - so that is why the unit tests included the phase "reboot". 

See https://github.com/saltstack-formulas/docker-formula/issues/203  and I think there was CentOS ticket somewhere too.

**Comment by noelmcloughlin** [comment]
Same issue in arch: https://github.com/saltstack-formulas/docker-formula/pull/229  We need to figure out if it is package or archive or both formats that encounter the issue.

**Comment by noelmcloughlin** [comment]
@myii @hatifnatt see https://github.com/saltstack-formulas/salter/issues/44#issue-484043138

**Comment by myii** [comment]
@noelmcloughlin OK, the `fedora-33` failure to start the service is due to this problem:

```sls
systemd[1]: /usr/lib/systemd/system/docker.service:16: PIDFile= references a path below legacy directory /var/run/, updating /var/run/docker.pid → /run/docker.pid; please update the unit file accordingly.
```

* I.e. Need to use `/run/` instead of `/var/run/`.

That will probably be the same for the other two failures as well.

**Comment by noelmcloughlin** [comment]
Yes that makes sense. @hatifnatt could you update the PR based on the test results please?

**Comment by myii** [comment]
> Yes that makes sense. @hatifnatt could you update the PR based on the test results please?

Don't worry, I'll push the fixed tests once I figure out the right ignores to use.

@hatifnatt Is `Allow edits from maintainers` selected?

**Comment by noelmcloughlin** [comment]
> dockerd[1548]: failed to start daemon: Error initializing network controller: error obtaining controller instance: failed to create NAT chain DOCKER: Iptables not found

Here is ugly hack: https://forums.docker.com/t/failing-to-start-dockerd-failed-to-create-nat-chain-docker/78269/9 
Yes - I think its best to disable those failing checks.

**Comment by myii** [comment]
OK, the other two are failing with the same.

`archive-fedora-32-master-py3`:

```sls
dockerd[803]: failed to start daemon: Error initializing network controller: error obtaining controller instance: failed to create NAT chain DOCKER: Iptables not found
```

`archive-opensuse-leap-152-master-py3`:

```sls
dockerd[758]: failed to start daemon: Error initializing network controller: error obtaining controller instance: failed to create NAT chain DOCKER: Iptables not found
```

So let me fix up those tests and then push them here.

**Comment by myii** [comment]
@noelmcloughlin @hatifnatt Tests updated in c168ee1, service is being tested wherever possible.

**Comment by hatifnatt** [comment]
@myii thanks for the fixes!

About `iptables` issues - `iptables` can be considered as mandatory and it's seems it can work with `firewalld` too
https://docs.docker.com/network/iptables/
I have already added `iptables` to `docker:pkg:deps` in #275 for CentOS 7 and 8 it was already here fo  RedHat `os_family`
https://github.com/saltstack-formulas/docker-formula/blob/3f8513119a5ce32dcb8413f983e582fea5defa28/docker/osfamilymap.yaml#L56
but due lists are not merged by `grains.filter_by` it's lost when `docker:pkg:deps` defined in `osfamilymap`

**Comment by hatifnatt** [comment]
@noelmcloughlin could you please review and give any conclusion?

**Comment by noelmcloughlin** [comment]
> could you please review and give any conclusion?

I checked and found https://github.com/saltstack-formulas/docker-formula/issues/273#issuecomment-797951343 and I was suggesting this line be fixed:  https://github.com/saltstack-formulas/docker-formula/blob/943ca5e9dbe5a6876181e8865fb308a68bf318eb/docker/software/archive/install.sls#L97

- Add `docker.pkg.docker.binary.name: docker` to `YAML` 
- Change line 97 to `start: {{ d.pkg.docker.path }}/{{ d.pkg.docker.service.name }}`

But I think you were not keen.   @myii  has approved feel free to merge yourself.

**Comment by hatifnatt** [comment]
I outlined my thoughts about additional variable in https://github.com/saltstack-formulas/docker-formula/issues/273#issuecomment-798787336 Do you still think additional variable is required? 
Docker daemon binary is named `dockerd` so even if add new variable it will be `docker.pkg.docker.binary.name: dockerd` if it's supposed to be used in
https://github.com/saltstack-formulas/docker-formula/blob/943ca5e9dbe5a6876181e8865fb308a68bf318eb/docker/software/archive/install.sls#L97
Actually I already wrote my own formula for docker, just want to close this 'hanging' PR in some way.

**Review by noelmcloughlin — APPROVED** [review]
Nice work.  You are probably correct.  I reviewed the CI and both package/archive states are working across an impressive matrix of OS.  So LGTM

**Comment by noelmcloughlin** [comment]
Because your PR includes this fix:  `start: {{ d.pkg.docker.path }}/dockerd`

**Comment by hatifnatt** [comment]
@noelmcloughlin thanks for review and merge!

**Pull request merged**
